src/main/io/gps.c

```diff
@@ -1333,21 +1345,25 @@ void gpsUpdate(timeUs_t currentTimeUs)
     gpsData.now = millis();
 
     if (gpsPort) {
+        const uint32_t initialCycleCount = getCycleCounter();
+
         DEBUG_SET(DEBUG_GPS_CONNECTION, 7, serialRxBytesWaiting(gpsPort));
         static uint8_t wait = 0;
         static bool isFast = false;
+
         while (serialRxBytesWaiting(gpsPort)) {
             wait = 0;
             if (!isFast) {
                 rescheduleTask(TASK_SELF, TASK_PERIOD_HZ(TASK_GPS_RATE_FAST));
                 isFast = true;
             }
-            if (cmpTimeUs(micros(), currentTimeUs) > GPS_RECV_TIME_MAX) {
+            if ((getCycleCounter() - initialCycleCount) > cpuCycleLimit()) {
```

Comment:
`cmp32()` here. Subtraction result is unsigned and it will fil on wraparound. And in this case, it may happen every few seconds.

Reply:
Is there some kind of Betaflight or MCU specific weirdness going on with unsigned int arithmetics? As far as I can tell the current code should handle wraparound as expected.

I did a small test to be sure
```C
unsigned int val1 = 2U;
unsigned int val2 = 4294967295U;  // 2^32 - 1
printf("value: %lld \n", val1 - val2);
```
which resulted in 
```
value: 3
```
as expected.

Reply:
The `CmpTimeUs` macro handles wrap. See `cmpTimeCycles` for cycle comparison. Used in the scheduler.

Reply:
I must be missing something here. Why should the result of the subtraction be converted to a signed value if it is known which value that is older? As far as I can tell the result will be the same in most cases, and if the difference is large it will result in a negative value, which we do not want here (then, if the difference ever become that large there are bigger problems elsewhere).

Reply:
@tbolin : Sorry, I was wrong. The wrap does work as intended.

> Why should the result of the subtraction be converted to a signed value _if it is known which value that is older_?

For ms/us, 31bits is considerably longer than any intended delay. Signed value does work independent of value ordering, so there is lesser chance to get it wrong. ( `cmp32(a, b) > tout` / `cmp32(a, b+tout) > 0`). Also, cmpTimeUs / cmp32 marks calculation with values that are expected to overflow (so `cmpTimeCyclesU(a, b)` instead of `a - b`).

For tick counter, the situation is more complicated due to range. Personally, I'd use signed conversion and limit maximum delay to some small value (<1s). This should be enforced (compile-time asserts / runtime saturation). Without this, code may behave differently depending on clock speed, leading to support nightmare. We will still run in problems with 2GHz chips, but that will take some time.

Reply:
BTW: It may be good idea to use

```
uint32_t getCycleCounter(void)
{
    return DWT->CYCCNT | 1;
}
```

This way, you can use `if (timeout && cmp32(getCycleCounter(), timeout) < 0)` safely and resolution loss is negligible.